plugins/soundsourcemediafoundation/audiosourcemediafoundation.cpp

```diff
@@ -0,0 +1,558 @@
+                m_leftoverBufferLength);
+        if (m_leftoverBufferLength > 0) {
+            if (framesNeeded != 0) {
+                qWarning() << __FILE__ << __LINE__
+                        << "WARNING: Expected frames needed to be 0. Abandoning this file.";
+                m_dead = true;
+            }
+            m_leftoverBufferPosition += frameCount;
+        }
+    } else {
+        // leftoverBuffer already empty or in the wrong position, clear it
+        m_leftoverBufferLength = 0;
+    }
+
+    while (!m_dead && framesNeeded > 0) {
+        HRESULT hr(S_OK);
+        DWORD dwFlags(0);
+        qint64 timestamp(0);
+        IMFSample *pSample(NULL);
+        bool error(false); // set to true to break after releasing
+
+        hr = m_pReader->ReadSample(MF_SOURCE_READER_FIRST_AUDIO_STREAM, // [in] DWORD dwStreamIndex,
+                0,                                 // [in] DWORD dwControlFlags,
+                NULL,                      // [out] DWORD *pdwActualStreamIndex,
+                &dwFlags,                        // [out] DWORD *pdwStreamFlags,
+                &timestamp,                     // [out] LONGLONG *pllTimestamp,
+                &pSample);                         // [out] IMFSample **ppSample
+        if (FAILED(hr)) {
+            qWarning() << "ReadSample failed!";
+            break; // abort
+        }
+
+        if (sDebug) {
+            qDebug() << "ReadSample timestamp:" << timestamp << "frame:"
+                    << frameFromMF(timestamp) << "dwflags:" << dwFlags;
+        }
+
+        if (dwFlags & MF_SOURCE_READERF_ERROR) {
+            // our source reader is now dead, according to the docs
+            qWarning()
+                    << "SSMF: ReadSample set ERROR, SourceReader is now dead";
+            m_dead = true;
+            break;
+        } else if (dwFlags & MF_SOURCE_READERF_ENDOFSTREAM) {
+            qDebug() << "SSMF: End of input file.";
+            break;
+        } else if (dwFlags & MF_SOURCE_READERF_CURRENTMEDIATYPECHANGED) {
+            qWarning() << "SSMF: Type change";
+            break;
+        } else if (pSample == NULL) {
+            // generally this will happen when dwFlags contains ENDOFSTREAM,
+            // so it'll be caught before now -bkgood
+            qWarning() << "SSMF: No sample";
+            continue;
+        } // we now own a ref to the instance at pSample
+
+        IMFMediaBuffer *pMBuffer(NULL);
+        // I know this does at least a memcopy and maybe a malloc, if we have
+        // xrun issues with this we might want to look into using
+        // IMFSample::GetBufferByIndex (although MS doesn't recommend this)
+        if (FAILED(hr = pSample->ConvertToContiguousBuffer(&pMBuffer))) {
+            error = true;
+            goto releaseSample;
+        }
+        sample_type *buffer(NULL);
+        DWORD bufferLengthInBytes(0);
+        hr = pMBuffer->Lock(reinterpret_cast<quint8**>(&buffer), NULL, &bufferLengthInBytes);
```

Comment:
@uklotzde: Are you fully understand this section? I not. If yes some line comments would help here.